api/turing/api/ensemblers_api.go

```diff
@@ -109,18 +111,221 @@ func (c EnsemblersController) UpdateEnsembler(
 			"Ensembler type cannot be changed after creation")
 	}
 
+	pyFuncEnsembler, isPyfunc := ensembler.(*models.PyFuncEnsembler)
+
+	var oldPyFuncEnsembler models.PyFuncEnsembler
+	if isPyfunc {
+		oldPyFuncEnsembler = models.PyFuncEnsembler{
+			RunID:       pyFuncEnsembler.RunID,
+			ArtifactURI: pyFuncEnsembler.ArtifactURI,
+		}
+	}
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
-		return InternalServerError("failed to update an ensembler", err.Error())
+		// Delete If Only RunID is changed
+		// If only the ensembler name that changed, mlflow won't create a new run, so we don't need to delete old / new run
+		if isPyfunc && oldPyFuncEnsembler.RunID != pyFuncEnsembler.RunID {
+			pyFuncErr := c.MlflowService.DeleteRun(context.Background(),
+				pyFuncEnsembler.RunID, pyFuncEnsembler.ArtifactURI, true)
+			if pyFuncErr != nil {
+				return InternalServerError("failed to update the ensembler", "cleanup process failed")
+			}
+		}
+		return InternalServerError("failed to update the ensembler", err.Error())
+	}
+
+	// Delete If Only RunID is changed
+	// If only the ensembler name that changed, mlflow won't create a new run, so we don't need to delete old / new run
+	if isPyfunc && oldPyFuncEnsembler.RunID != pyFuncEnsembler.RunID {
+		err = c.MlflowService.DeleteRun(context.Background(), oldPyFuncEnsembler.RunID, oldPyFuncEnsembler.ArtifactURI, true)
+		if err != nil {
+			fmt.Printf("failed to delete mlflow old run: %s", err.Error())
+		}
 	}
 
 	return Ok(ensembler)
 }
 
+func (c EnsemblersController) DeleteEnsembler(
+	_ *http.Request,
+	vars RequestVars,
+	_ interface{},
+) *Response {
+	options := EnsemblersPathOptions{}
+
+	if err := c.ParseVars(&options, vars); err != nil {
+		return BadRequest("failed to fetch ensembler",
+			fmt.Sprintf("failed to parse query string: %s", err))
+	}
+
+	ensembler, err := c.EnsemblersService.FindByID(
+		*options.EnsemblerID,
+		service.EnsemblersFindByIDOptions{
+			ProjectID: options.ProjectID,
+		})
+
+	if err != nil {
+		return NotFound("ensembler not found", err.Error())
+	}
+
+	// First we need to check if there are active job / router using the ensembler
+	// If such ensembler exist, the deletion process are restricted
+	// Check if there are any deployed / pending router version using the ensembler
+	httpStatus, err := c.checkActiveRouterVersion(options)
```

Comment:
i dont think we need to put that to check again, since deleting the router version that is the current version already prohibited byt the deletion itself (There are already existing check in the delete router version (we cant delete current router version)